### ISSUE TITLE
Mix functional and classical gradient computations

### ISSUE THREAD

**Issue by RPegoud** [RPegoud/Hercules#5]
The previous issue (#3 )  was created to convert the functional gradient computation to the usual pytorch approach using `backward()`. This was meant as a fix to enable `accelerate` to use multiple gpus in parallel.
However this new approach leads to bad performances since we do not make use of momentum nor the forget gate anymore, which causes the model to "overfit" by perpetually recording information without forgetting. This pollutes the transformer's outputs and leads to an increasing causal loss.

To fix these issues, I will try to:
- **Revert back to the old functional paradigm**, using learned momentum, learning rate and forget gates. The neural memory will be trained manually at inference time, **however the gates will be trained using** `backward()`. This will give `accelerate` gradients to compute and sync.
- Implement the **memory as a gate** architecture, which will enable the model to ignore the memory outputs if needed. This should prevent the causal loss from increasing too much if memory is unnecessary for the current inputs.
